Ad sgima delta fixes

tachicialex opened the pull request:
This series of patches fix a couple of issues in ad_sigma_delta layer that have been observed when testing the ad7124.

commodo commented on drivers/iio/adc/ad_sigma_delta.c:
```diff
@@ -435,6 +436,8 @@ static int ad_sd_buffer_postenable(struct iio_dev *indio_dev)
 		spi_engine_offload_enable(sigma_delta->spi, true);
 	} else {
 		sigma_delta->spi_transfer[1].rx_buf = sigma_delta->buf_data;
+		reg_size = sigma_delta->spi_transfer[1].len;
```
just to make sure here;
this also works on other sigma-delta drivers, right?

when sending this patch upstream, you want to be careful to not mention *just* ad7124;
you'd want to make sure that all Sigma-Delta drivers work with this change;

commodo commented on drivers/iio/adc/ad_sigma_delta.c:
```diff
@@ -291,6 +291,23 @@ static int ad_sigma_delta_set_active_slots(struct ad_sigma_delta *sigma_delta,
 	return 0;
 }
 
+static int ad_sigma_delta_check_seq_sync(struct ad_sigma_delta *sigma_delta)
```
this patch `iio: adc: ad_sigma_delta: add sequencer sync` looks like is doing a bit too much in one go;
i'll take a look at this later;
one idea would be to also split it a bit;

tachicialex commented:
Changelog V2:
- changed commit "ad_sigma_delta: Sync with chip sequencer " description to explain why this patch fixes a bug in ad_sigma_delta
- added a cleanup commit

commodo commented on include/linux/iio/adc/ad_sigma_delta.h:
```diff
@@ -88,9 +91,12 @@ struct ad_sigma_delta {
 	unsigned int		active_slots;
 	unsigned int		current_slot;
+	unsigned int		converted_chan;
+	unsigned int		sequencer_desync;
```
the `unsigned int		converted_chan;` does not look like is used anywhere;
remove it?

also looks like you could drop `unsigned int		sequencer_desync;` as well;

commodo commented on drivers/iio/adc/ad_sigma_delta.c:
```diff
@@ -480,24 +508,42 @@ static irqreturn_t ad_sd_trigger_handler(int irq, void *p)
+	if (info->get_converted_channel)
+		chan_num = ad_sigma_delta_check_seq_sync(sigma_delta);
+
+	if (sigma_delta->sequencer_desync) {
```
it looks to me that if you move this if-block inside `ad_sigma_delta_check_seq_sync()` and converge the logic somewhere in there, then `sigma_delta->sequencer_desync` is not needed;

tachicialex commented:
Changelog V3:
- removed desync related commits

mhennerich commented on drivers/iio/adc/ad_sigma_delta.c:
```diff
@@ -435,6 +436,8 @@ static int ad_sd_buffer_postenable(struct iio_dev *indio_dev)
 	} else {
 		sigma_delta->spi_transfer[1].rx_buf = sigma_delta->buf_data;
+		reg_size = sigma_delta->spi_transfer[1].len;
+		sigma_delta->spi_transfer[1].rx_buf += 4 - reg_size;
```
Is it guaranteed that reg_site can never become > 4? 
If that happens we access non allocated memory space.

tachicialex replied:
The spi transfer is currently 3 bytes on all sigma-delta chips because all sigma-delta chips have max 3 bytes data-registers. In the future the spi transfer can become 4 bytes if we decide to add CRCs appended at the end to each transfer.

mhennerich replied:
Thanks, BUG_ON(reg_size > 4) wouldn't harm.

tachicialex replied:
Ok, I will come with another pr to add that check.
